src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java

```diff
@@ -968,353 +1432,197 @@ public String toString() {
-         */
-        public ShutdownOnSuccess(String name, ThreadFactory factory) {
-            super(name, factory);
+        // list of forked subtasks, only accessed by owner thread
+        private final List<Subtask<T>> subtasks = new ArrayList<>();
+
+        @Override
+        public boolean onFork(Subtask<? extends T> subtask) {
+            @SuppressWarnings("unchecked")
+            var tmp = (Subtask<T>) Objects.requireNonNull(subtask);
+            subtasks.add(tmp);
+            return false;
         }
 
-        /**
-         * Constructs a new unnamed {@code ShutdownOnSuccess} that creates virtual threads.
-         *
-         * @implSpec This constructor is equivalent to invoking the 2-arg constructor with
-         * a name of {@code null} and a thread factory that creates virtual threads.
-         */
-        public ShutdownOnSuccess() {
-            this(null, Thread.ofVirtual().factory());
+        @Override
+        public boolean onComplete(Subtask<? extends T> subtask) {
+            return (subtask.state() == Subtask.State.FAILED)
+                    && (firstException == null)
+                    && FIRST_EXCEPTION.compareAndSet(this, null, subtask.exception());
         }
 
         @Override
-        protected void handleComplete(Subtask<? extends T> subtask) {
-            if (firstResult != null) {
-                // already captured a result
-                return;
+        public Stream<Subtask<T>> result() throws Throwable {
+            Throwable ex = firstException;
+            if (ex != null) {
+                throw ex;
+            } else {
+                return subtasks.stream();
             }
+        }
+    }
 
-            if (subtask.state() == Subtask.State.SUCCESS) {
-                // task succeeded
-                T result = subtask.get();
-                Object r = (result != null) ? result : RESULT_NULL;
-                if (FIRST_RESULT.compareAndSet(this, null, r)) {
-                    super.shutdown();
-                }
-            } else if (firstException == null) {
-                // capture the exception thrown by the first subtask that failed
-                FIRST_EXCEPTION.compareAndSet(this, null, subtask.exception());
-            }
+    /**
+     * A joiner that returns the result of the first subtask to complete successfully.
+     * Cancels the scope if any subtasks succeeds.
+     */
+    private static final class AnySuccessful<T> implements Joiner<T, T> {
+        private static final VarHandle FIRST_SUCCESS;
+        private static final VarHandle FIRST_EXCEPTION;
+        static {
+            MethodHandles.Lookup l = MethodHandles.lookup();
+            FIRST_SUCCESS = MhUtil.findVarHandle(l, "firstSuccess", Subtask.class);
+            FIRST_EXCEPTION = MhUtil.findVarHandle(l, "firstException", Throwable.class);
         }
+        private volatile Subtask<T> firstSuccess;
+        private volatile Throwable firstException;
```

Comment:
Not strictly needed, but it would be possible to have a single `Object`-typed field called `first` here which is either a Subtask in case of first success, and a Throwable in case of first failure, which could simplify the tracking.

```
switch(first) {
    case Subtask<?> s -> already done
    case Throwable t -> swap out `first` if the next task is a success
    default -> cas first expecting null with the given subtask
}
```